### PR TITLE
Sgrasse/develop/1526 refactoring

### LINKED ISSUE #1526: validation_io can be simplified
### Describe the task. It can be a feature, documentation, etc.

validation_io is currently unnecessarily complex since it converts the new data structures for ModelOutput etc to the old list-of-lists based one to be compatible with weathgen.common.io. The latter should be adapated to directly use the new, dict-based data structures.

CC @grassesi 

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by grassesi**
## Issue Number

closes #1526 
Additional improvements and cleanups that do not change existing behavior, this is complementary to #1908 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/datasets/batch.py [hunk 1]
```diff
@@ -295,10 +299,15 @@ def __init__(
             streams, num_target_samples, output_steps, self.output_idxs
         )
 
-        self.source2target_matching_idxs = np.full(num_source_samples, -1, dtype=np.int32)
-        self.target2source_matching_idxs = [[] for _ in range(num_target_samples)]
+        self.source2target_matching_idxs: np.typing.NDArray[np.int32] = np.full(
+            num_source_samples, -1, dtype=np.int32
+        )
+        self.target2source_matching_idxs: np.typing.NDArray[np.int32] = [
+            [] for _ in range(num_target_samples)
```
Comment (TillHae) [inline src/weathergen/datasets/batch.py:306]: This doesn't look like a numpy array to me and rather like a Python list of empty lists

File: src/weathergen/datasets/batch.py
Comment (TillHae) [inline src/weathergen/datasets/batch.py:161]: In the `ModelBatch` class you apply `torch.device()`, but here you don't. It's probably not necessary in both cases, but we should keep it consistent.

File: src/weathergen/datasets/masking.py [hunk 2]
```diff
@@ -440,7 +440,7 @@ def build_samples_for_stream(
                 source_target_mapping += [target_idx]
                 i_source += 1
 
-        source_target_mapping = np.array(source_target_mapping, dtype=np.int32)
+        source_target_mapping: NDArray[np.int32] = np.array(source_target_mapping, dtype=np.int32)
```
Comment (TillHae) [inline src/weathergen/datasets/masking.py:443]: Up to this point you have only used `np.typing.NDArray[]`

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 3]
```diff
@@ -607,15 +610,16 @@ def _get_source_target_masks(self, training_mode):
         """
         Generate source and target masks for all streams.
         """
-        masks = {}
+
+        masks: dict[str, tuple[MaskData, MaskData, np.NDArray[np.int32]]] = {}
```
Comment (TillHae) [inline src/weathergen/datasets/multi_stream_data_sampler.py:614]: I think this has to be wrong. `np.NDArray` shouldn't exist. Either directly import `NDArray` or use `np.typing.NDArray`

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 4]
```diff
@@ -677,6 +684,7 @@ def _get_batch(self, idx: int, num_forecast_steps: int):
         for stream_info, (stream_name, stream_ds) in zip(
             self.streams, self.streams_datasets.items(), strict=True
         ):
+            stream_info: Config
```
Comment (TillHae) [inline src/weathergen/datasets/multi_stream_data_sampler.py:687]: My personal preference would be to move this above the for loop to keep everything inside the loop as readable as possible.

File: src/weathergen/datasets/stream_data.py [hunk 5]
```diff
@@ -198,51 +198,6 @@ def add_source(
 
         self.source_is_spoof[step] = ss_raw.is_spoof
 
-    def add_target(
```
Comment (TillHae) [inline src/weathergen/datasets/stream_data.py:201]: good catch 👍

File: src/weathergen/datasets/tokenizer_masking.py [hunk 6]
```diff
@@ -53,27 +54,36 @@ def reset_rng(self, rng) -> None:
...
-        hl = self.healpix_level
-        token_size = stream_info["token_size"]
-
-        tokens = []
+        tokens: list[tuple[list[list[torch.Tensor | None]], list[list[int]]]] = []
+        do_tokenize_spacetime = stream_info.get("tokenize_spacetime", False)
         for rdata in data:
             # skip empty data
             if rdata.is_empty():
                 tokens += [(None, None)]
```
Comment (TillHae) [inline src/weathergen/datasets/tokenizer_masking.py:69]: We should replace this with `tokens.append((None, None))`

File: src/weathergen/datasets/tokenizer_masking.py [hunk 7]
```diff
@@ -53,27 +54,36 @@ def reset_rng(self, rng) -> None:
...
+                )
+            else:
+                idxs_cells, idxs_cells_lens = tokenize_space(
+                    readerdata_to_torch(rdata),
+                    stream_info["token_size"],
+                    self.healpix_level,
+                    pad_tokens,
+                )
+
             tokens += [(idxs_cells, idxs_cells_lens)]
```
Comment (TillHae) [inline src/weathergen/datasets/tokenizer_masking.py:87]: We should replace this with `tokens.append((idxs_cells, idxs_cells_lens))`

File: src/weathergen/datasets/tokenizer_masking.py
Comment (TillHae) [inline src/weathergen/datasets/tokenizer_masking.py:96]: for `masker.build_samples_for_stream` you updated the return type to `tuple[MaskData, MaskData, np.typing.NDArray[np.int32]]`, but here it still uses the old one.

File: src/weathergen/datasets/tokenizer_utils.py [hunk 8]
```diff
@@ -199,22 +202,27 @@ def tokenize_spacetime(
...
 
     num_healpix_cells = 12 * 4**hl
-    idxs_cells = [[] for _ in range(num_healpix_cells)]
-    idxs_cells_lens = [[] for _ in range(num_healpix_cells)]
+    idxs_cells: list[list[Tensor | None]] = [[] for _ in range(num_healpix_cells)]
+    idxs_cells_lens: list[list[int]] = [[] for _ in range(num_healpix_cells)]
 
     offset_step = 0
-    t_unique = np.unique(rdata.datetimes)
+    t_unique: typing.iterable[NPDT64] = np.unique(rdata.datetimes)
```
Comment (TillHae) [inline src/weathergen/datasets/tokenizer_utils.py:209]: `typing.Iterable[NPDT64]` is the correct one

File: src/weathergen/datasets/tokenizer_utils.py [hunk 9]
```diff
@@ -199,22 +202,27 @@ def tokenize_spacetime(
...
     for _, t in enumerate(t_unique):
+        t: NPDT64
```
Comment (TillHae) [inline src/weathergen/datasets/tokenizer_utils.py:211]: This is not needed as it is already given through the last type hint

File: src/weathergen/datasets/tokenizer_utils.py [hunk 10]
```diff
@@ -199,22 +202,27 @@ def tokenize_spacetime(
...
 
-        # collect data for all time steps
-        idxs_cells = [t + tc for t, tc in zip(idxs_cells, idxs_cur, strict=True)]
-        idxs_cells_lens = [t + tc_l for t, tc_l in zip(idxs_cells_lens, idxs_cur_lens, strict=True)]
+        # append tokens/n_tokens for current time step for each healpix cell to existing tokens
+        for cell_steps_tokens, cell_steps_tokens_cur, cell_steps_lens, cell_steps_lens_cur in zip(
+            idxs_cells, idxs_cur, idxs_cells_lens, idxs_cur_lens, strict=True
+        ):
+            cell_steps_tokens.extend(cell_steps_tokens_cur)
+            cell_steps_lens.extend(cell_steps_lens_cur)
```
Comment (TillHae) [inline src/weathergen/datasets/tokenizer_utils.py:224]: I like this a lot 👍

File: src/weathergen/model/model.py
Comment (TillHae) [inline src/weathergen/model/model.py:62]: If `physical: list[dict[StreamName, tuple[torch.Tensor]]]` is correct
then `pred: torch.Tensor` has to be wrong

File: src/weathergen/train/loss_modules/loss_module_physical.py
Comment (TillHae) [inline src/weathergen/train/loss_modules/loss_module_physical.py:190]: This argument doesn't exist

File: src/weathergen/train/target_and_aux_module_base.py [hunk 11]
```diff
@@ -9,32 +9,29 @@
...
+from weathergen.datasets.batch import SampleMetaData
 from weathergen.model.engines import LatentState
 
 type StreamName = str
 
 
-@dataclasses.dataclass
 class TargetAuxOutput:
     """
     A dataclass to encapsulate the TargetAndAuxCalculator output and give a clear API.
```
Comment (TillHae) [inline src/weathergen/train/target_and_aux_module_base.py:22]: This is no longer a dataclass

File: src/weathergen/train/target_and_aux_module_base.py [hunk 12]
```diff
@@ -9,32 +9,29 @@
...
+            dict[
+                StreamName,
+                dict[str, list[torch.Tensor] | list[dict[str, SampleMetaData]] | list[bool]],
+            ]
+        ] = [{} for _ in range(len_target)]
+        self.latent: list[dict[str, torch.Tensor | LatentState]] = [{} for _ in range(len_target)]
+        self.aux_outputs: dict[str, torch.Tensor] = {}
 
     def add_physical_target(
         self, timestep_idx: int, stream_name: StreamName, pred: torch.Tensor
```
Comment (TillHae) [inline src/weathergen/train/target_and_aux_module_base.py:37]: I think `pred` is a dictionary with tensors as its values as you also pointed out in the constructor.

File: src/weathergen/train/trainer.py [hunk 13]
```diff
@@ -11,6 +11,7 @@
 import copy
 import logging
 import time
+from collections.abc import Iterable
```
Comment (TillHae) [inline src/weathergen/train/trainer.py:14]: Before you used typing.Iterable

File: src/weathergen/train/trainer.py [hunk 14]
```diff
@@ -77,8 +80,8 @@ def __init__(self, train_logging: Config):
         self.model_params = None
         self.optimizer: torch.optim.Optimizer | None = None
         self.t_start: float = 0
-        self.target_and_aux_calculators = None
-        self.target_and_aux_calculators_val = None
+        self.target_and_aux_calculators: dict[str, TargetAndAuxModuleBase] = None
+        self.target_and_aux_calculators_val: dict[str, TargetAndAuxModuleBase] = None
```
Comment (TillHae) [inline src/weathergen/train/trainer.py:84]: Therefore those two arguments can also be None and not just a dict

**Review by TillHae — CHANGES_REQUESTED** [review]
